File: Tests/SwitchThemesNXTests/Layouts.cpp
#include "CppUnitTest.h"
#include "../../SwitchThemesNX/source/SwitchThemesCommon/SwitchThemesCommon.hpp"
#include "../../SwitchThemesNX/source/SwitchThemesCommon/Layouts/Patches.hpp"
#include "Util.hpp"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace SwitchThemesNXTests
{
	TEST_CLASS(Layouts)
	{
	public:
		TEST_METHOD(LayoutLoading)
		{
			// The theme injector will remove useless fields from jsons to save space, this test ensures the jsons keep the same meaning for the C++ parser
			// The ParsedLayouts folder is populated by SwitchThemesCommonTests.Layouts.LoadAndOptimizeAll
			for (const auto& f : std::filesystem::directory_iterator("../SwitchThemes/layouts/"))
			{
				if (f.is_regular_file() && f.path().extension() == ".json")
				{
					auto a = Patches::LoadLayout(Util::ReadAllText(f.path().string()));
					auto b = Patches::LoadLayout(Util::ReadTestString("ParsedLayouts/" + f.path().filename().string()));

					if (a != b)
						throw std::runtime_error("Files don't match: " + f.path().filename().string());
				}				
			}
		}
	};
}

File: Tests/SwitchThemesCommonTests/Layouts.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SwitchThemes.Common;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace SwitchThemesCommonTests
{
	[TestClass]
	public class Layouts
	{
		[TestMethod]
		public void LoadAndOptimizeAll() 
		{
			// This generates layouts to load for SwitchThemesNXTests::Layouts::LayoutLoading
			foreach (var f in Directory.EnumerateFiles("../../../../../SwitchThemes/layouts").Where(x => x.EndsWith(".json")))
				File.WriteAllText(
					Path.Combine(Util.GetPath("ParsedLayouts"), Path.GetFileName(f)), 
					LayoutPatch.Load(File.ReadAllText(f)).AsJson()
				);
		}
		
	}
}
